### ISSUE TITLE
Behavior of `values` getter in enhanced enums.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1994]
The existing `enum` declaration introduces a `values` static constant with a list of the enum elements.

The [current proposal](https://github.com/dart-lang/language/blob/master/accepted/future-releases/enhanced-enums/feature-specification.md) for enhanced enums states that it also introduces such a static constant, and it's a compile-time if that would conflict with an inherited or declared member or constructor named `values` (or any other conflict, really).

This differs from how the `toString` method is handled.
If the enhanced enum declaration overrides `Object.toString`, then no backwards compatible `toString` is added to the `enum` class.

Should we instead say that:
* If the `enum` declaration inherits or declares a member or constructor with base name `values`, then no new `values` static constant is introduced.
* Otherwise a declaration of `static const List<EnumName> values = [v1, v2, ..., vn];` is introduced, containing all the enum elements.

The type of the list is the raw type of the enum class type, `EnumName`. If the `enum` declaration is generic, and it's not possible to find a valid, well-bounded or super-bounded, type instantiation using instantiate-to-bounds, then it's a compile-time error.
In that case, the author can write their own `static const` `values` declaration with a valid bound, if *they* can find one.

If not, they may want to declare a `values` member anyway, to avoid the default one which will be a compile-time error.
That's a little icky, having to introduce an otherwise not-usable declaration to avoid a default declaration which is a compile-time error.

Another alternative is to also omit the default declaration if it cannot be typed. 
If anyone hits that (by creating an `enum` with generic type parameters which has no instantiate to bounds, then they might be confused by there not being any `values` list.

Most likely that will never happen in practice, and it won't matter which approach we use.

**Comment by munificent** [comment]
> Should we instead say that:
> 
> * If the `enum` declaration inherits or declares a member or constructor with base name `values`, then no new `values` static constant is introduced.
> * Otherwise a declaration of `static const List<EnumName> values = [v1, v2, ..., vn];` is introduced, containing all the enum elements.

Yes, I like this. Consistency with how other members are automatically synthesized seems good. I think this could be practically useful too. I wouldn't be entirely surprised to see users define an enum with a few extra "hidden" values that they don't want to include in `values` but are technically part of the enum. This would give them a way to write that.

**Comment by lrhn** [comment]
Having hidden values has the risk (read: guarantee) of triggering "switch non-exhaustive" warnings.

Introducing a new value to an enum is a breaking change if anyone relies on exhaustiveness in the control flow analysis.
Having a private value in an enum prevents anyone else from actually having exhaustive switches, they always need to consider a `default` or fall-through case. That's an invisible side-effect of having a private member.
Then it also becomes non-breaking to add extra elements.
I guess having an enum with a hidden `_dummy` element, which is never leaked, to *prevent* exhaustive switches could be a strategy to allow the enum to be growable in the future. (Not great design, but it'll work, and there is no other way to disable exhaustiveness checks, other than not use an `enum` to begin with.)

We could define enum-exhaustiveness to only count *accessible* enum values.
That's still not sufficient, since you *can* leak the private values, and then a switch which returns on every accessible enum value can still pass through without returning. Being exhaustive is then only a requirement and constraint, not a help in any way.

We could disable exhaustiveness checks for enums with inaccessible elements, so at least we *won't* complain about a switch *not* being exhaustive, and force you to write a `default: /*do nothing*/`.
Or we could still warn if you don't exhaustively check the accessible elements, and then allow fall-thorough or a `default:` case, and expect them to be possible in flow analysis.

We could disallow private enum elements entirely. It's not entirely unreasonable, it's an important part of the enum's API and it's invisible outside the current library (like we disallow optional named parameters with private names, even though it's only really a problem for interface member overriding, static methods and function values have no problems at all).

Or we could just keep doing what we're doing now and allow enums that it's impossible to be exhaustive about, and not worry since nobody actually makes them. (Unless this change changes people's behavior, and they start doing it because now they can).

**Comment by Levi-Lesches** [comment]
I don't think Dart should offer a way for authors to disable their users' exhaustiveness checking. Whether the enum officially "supports" the idea of exhaustiveness is up to interpretation, and whether a future expansion should trigger errors is a matter of preference. Disabling exhaustiveness checking to prevent breaking changes gets at the wrong point of the issue: errors caused by breaking changes are annoying, sure, but they're helpful to show users what they need to do, and there're always `// ignore` comments for those who don't want the help.

**Comment by lrhn** [comment]
Adding an inaccessible enum value *prevents* being exhaustive, which means that you either have to add a `default:` clause or an `// ignore:` comment to avoid a warning. At that point you have turned off exhaustiveness checking *anyway* (having a potentially reachable `default:` clause in an otherwise non-exhaustive switch does that). 

So, we have already given authors a way to force users to turn off exhaustiveness checking. It's just not very convenient.

We can't *currently* prevent the private enum value from leaking, because it's in `values`, which can always be accessed if the type itself is accessible. With this proposal we can also do that.

**Comment by eernstg** [comment]
> Should we instead say that:
> 
> * If the `enum` declaration inherits or declares a member or constructor with base name `values`, then no new `values` static constant is introduced.
> * Otherwise a declaration of `static const List<EnumName> values = [v1, v2, ..., vn];` is introduced, containing all the enum elements.

I agree with @munificent that this is a meaningful and reasonably consistent approach.

With respect to exhaustiveness, I'd prefer an approach where an explicitly declared `values` always disables exhaustiveness (even if it would work, say, because the declaration is identical to the one that we would have generated). It should be flagged in `switch` statements without a `default`: Even in the case where we could argue that this enum was _designed_ to consider a subset of the declared values as exhaustive, the correctness danger associated with unsound exhaustiveness is greater than the inconvenience of having to write that `default` clause, even in the case where the intention is that the default clause cannot do anything other than `default: throw "The library that declares MyEnum is buggy!";`

Same thing with private values: Allow them, and mark switches in other libraries as non-exhaustive.

After all, exhaustiveness is not relevant for comparisons like `myEnumValue == MyEnum.aSpecificValue` or for passing actual arguments, and exhaustiveness is not supported for maps or sets (yes, it could be useful for maps with constant keys ;-), so there are lots of usages of enum values where the lack of exhaustiveness is not a problem.

**Comment by lrhn** [comment]
You can't declare a `values`.
